JSONObject.java

```diff
@@ -1465,51 +1461,84 @@ public static Writer quote(String string, Writer w) throws IOException {
         String hhhh;
         int i;
         int len = string.length();
+        int prev = 0;
 
-        w.write('"');
+        w.append('"');
         for (i = 0; i < len; i += 1) {
             b = c;
             c = string.charAt(i);
             switch (c) {
             case '\\':
             case '"':
-                w.write('\\');
-                w.write(c);
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append('\\');
+                prev = i;
                 break;
             case '/':
                 if (b == '<') {
-                    w.write('\\');
+                    if(prev < i) {
+                        w.append(string, prev, i);
+                    }
+                    w.append('\\');
+                    prev = i;
                 }
-                w.write(c);
                 break;
             case '\b':
-                w.write("\\b");
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append("\\b");
+                prev = i + 1;
                 break;
             case '\t':
-                w.write("\\t");
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append("\\t");
+                prev = i + 1;
                 break;
             case '\n':
-                w.write("\\n");
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append("\\n");
+                prev = i + 1;
                 break;
             case '\f':
-                w.write("\\f");
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append("\\f");
+                prev = i + 1;
                 break;
             case '\r':
-                w.write("\\r");
+                if(prev < i) {
+                    w.append(string, prev, i);
+                }
+                w.append("\\r");
+                prev = i + 1;
                 break;
             default:
                 if (c < ' ' || (c >= '\u0080' && c < '\u00a0')
                         || (c >= '\u2000' && c < '\u2100')) {
-                    w.write("\\u");
+                    if(prev < i) {
+                        w.append(string, prev, i);
+                    }
                     hhhh = Integer.toHexString(c);
-                    w.write("0000", 0, 4 - hhhh.length());
-                    w.write(hhhh);
-                } else {
-                    w.write(c);
+                    w.append("\\u0000", 0, 6 - hhhh.length());
+                    w.append(hhhh);
+                    prev = i + 1;
                 }
             }
         }
-        w.write('"');
+        if(prev == 0) {
+            w.append(string);
```

Comment:
For the degenerate case of writing sequences that require no escaping, writing the entire CharSequence to the Appendable means the fewest temporary copies. Situations where this could be important include base-64 encoded binary data.